### PR TITLE
PIPE-1120 : Artifact CLI/SDK implementation

### PR DESCRIPTION
**Pull request by ydixit-clarifai**
## 📋 **Overview**

This PR implements a comprehensive artifact management system for the Clarifai Python SDK, providing both programmatic SDK access and CLI commands for managing artifacts and their versions. The implementation follows established project patterns and includes complete test coverage with robust error handling and user experience improvements.

#### **Artifact Management (`clarifai.client.artifact.Artifact`)**
- **CRUD Operations**: Create, read, update, delete artifacts
- **Listing & Search**: List artifacts with pagination support  
- **Metadata Management**: Handle artifact descriptions and metadata
- **Existence Checking**: Verify artifact existence before operations
- **Error Handling**: Comprehensive gRPC error handling with user-friendly messages
- **Mandatory Parameters**: All operations require explicit artifact_id, user_id, and app_id

#### **Artifact Version Management (`clarifai.client.artifact_version.ArtifactVersion`)**
- **File Upload/Download**: Streaming file operations with real-time progress tracking
- **Version Lifecycle**: Create, delete, and manage artifact versions
- **Chunked Transfer**: 14MB fixed chunk size
- **Retry Logic**: Robust retry mechanism with exponential backoff for network failures
- **Progress Tracking**: Real-time upload/download progress with tqdm integration
- **Resume Support**: Download resume capability for interrupted transfers
- **Type Safety**: Full type hints and comprehensive validation throughout
- **Visibility Management**: Support for private, public, and organization visibility settings

#### **Command Structure**
```bash
clarifai artifact [OPTIONS] COMMAND [ARGS]...
clarifai af [OPTIONS] COMMAND [ARGS]...        # Short alias support
```

#### **Available Commands**
- **`list/ls`**: List artifacts or artifact versions with formatted table output
- **`get`**: Get detailed artifact or artifact version information
- **`cp`**: Upload/download with automatic source/destination detection
- **`delete`**: Delete artifacts or versions with interactive confirmation prompts

#### **Command Examples**
```bash
# List artifacts
clarifai artifact list users/user123/apps/app456
clarifai af ls users/user123/apps/app456

# Upload files
clarifai artifact cp ./model.pt users/user123/apps/app456/artifacts/my-model
clarifai af cp ./weights.safetensors users/u/apps/a/artifacts/model --description="Version 2.0"

# Download files  
clarifai artifact cp users/user123/apps/app456/artifacts/my-model ./downloaded-model.pt
clarifai af cp users/u/apps/a/artifacts/model/versions/v123 ./specific-version.pt

# Get information
clarifai artifact get users/user123/apps/app456/artifacts/my-model
clarifai af get users/u/apps/a/artifacts/model/versions/v123

# Delete with confirmation
clarifai artifact delete users/user123/apps/app456/artifacts/my-model
clarifai af delete users/u/apps/a/artifacts/model/versions/v123 --force
```

### **Client Separation Pattern**
- **`Artifact`**: Metadata operations (create, delete, info, list, exists)
- **`ArtifactVersion`**: File operations (upload, download, version management)
- **Clean Separation**: Each client handles its specific domain responsibilities
- **Mandatory Parameters**: Consistent requirement for artifact_id, user_id, and app_id across operations

### **Error Handling Strategy**
- **gRPC Error Wrapping**: Converts gRPC errors to user-friendly `UserError` messages
- **Input Validation**: Comprehensive parameter validation with clear error messages
- **Network Resilience**: Retry logic with exponential backoff for transient network failures
- **User Experience**: Clear, actionable error messages for common scenarios
- **Progress State Reset**: Proper state reset between retry attempts for accurate progress tracking

### **Path Parsing & Validation**
- **Artifact Path Format**: `users/{user_id}/apps/{app_id}/artifacts/{artifact_id}`
- **Version Path Format**: `users/{user_id}/apps/{app_id}/artifacts/{artifact_id}/versions/{version_id}`
- **Local Path Detection**: Automatic detection of local vs remote paths
- **Validation**: Comprehensive path format validation with clear error messages

## 📁 **File Structure**

```
clarifai/
├── client/
│   ├── artifact.py              # Artifact metadata client
│   └── artifact_version.py      # Artifact version file operations client
├── cli/
│   └── artifact.py              # CLI commands with aliases and path parsing
└── constants/
    └── artifact.py             # Configuration constants and settings

tests/
├── test_artifact.py             # Artifact client tests (13 tests)
├── test_artifact_version.py     # ArtifactVersion client tests (23 tests)  
├── test_artifact_utils.py       # Shared test utilities and fixtures
├── cli/
│   ├── test_artifact_cli.py     # CLI command tests (30 tests)
│   └── test_artifact_version_cli.py # CLI version-specific tests (14 tests)
```

### **Comprehensive Test Coverage**
- **80+ Test Methods** across 4 main test files
- **11 Test Classes** organized by functionality
- **Unit Tests**: Individual component testing with comprehensive mocking
- **Integration Tests**: Full workflow validation and error handling
- **CLI Tests**: Command-line interface validation with real command simulation
- **Error Scenario Tests**: Edge cases, network failures, and user input validation

### **Test Organization & Improvements**
```
tests/
├── test_artifact.py              # 13 tests for Artifact client operations
├── test_artifact_version.py      # 23 tests for ArtifactVersion client operations  
├── test_artifact_utils.py        # Shared test utilities (eliminates code duplication)
├── cli/
│   ├── test_artifact_cli.py      # 30 tests for CLI commands and path parsing
│   └── test_artifact_version_cli.py # 14 tests for version-specific CLI operations
```

### **Upload/Download Specifications**
- **Chunk Size**: 14MB fixed chunks
- **Streaming**: Memory-efficient streaming for large files with tqdm progress bars
- **Progress Tracking**: Real-time progress indicators with accurate byte counting
- **Retry Logic**: Exponential backoff for network failures with state reset between attempts
- **Resume Support**: Download resume capability for interrupted transfers
- **Visibility Options**: Support for private, public, and organization-level visibility

### **Progress Tracking Improvements**
- **Accurate Indexing**: Separate tracking for config vs data chunks to prevent off-by-one errors
- **State Reset**: Proper progress state reset between retry attempts
- **Memory Efficiency**: Pre-calculated chunk sizes for accurate progress reporting
- **Error Recovery**: Graceful handling of unexpected responses during streaming

### **CLI Alias Support**
- **`artifact`** → **`af`**: Short alias for artifact command
- **`list`** → **`ls`**: Short alias for list subcommand
- **AliasedGroup**: Custom Click group supporting multiple command aliases

### **Upload/Download Specifications**
- **Chunk Size**: 14MB fixed chunks
- **Streaming**: Memory-efficient streaming for large files
- **Progress Tracking**: Real-time progress indicators
- **Retry Logic**: Exponential backoff for network failures
- **Compression**: Automatic handling of file compression

### **Path Format Specification**
```
# Artifact path format
users/{user_id}/apps/{app_id}/artifacts/{artifact_id}

# Artifact version path format  
users/{user_id}/apps/{app_id}/artifacts/{artifact_id}/versions/{version_id}

# Local path detection
./relative/path, /absolute/path, file.txt, ../parent/file
```

### **CLI Alias Support**
- **`artifact`** → **`af`**: Short alias for artifact command
- **`list`** → **`ls`**: Short alias for list subcommand
- **AliasedGroup**: Custom Click group supporting multiple aliases

### **Network Resilience**
- **Retry Logic**: Automatic retry with exponential backoff (up to 3 attempts by default)
- **Progress State Management**: Proper state reset between retry attempts to prevent UI confusion
- **Timeout Handling**: Configurable timeouts for different operations
- **Connection Management**: Efficient gRPC connection handling
- **Resume Support**: Download resume capability for interrupted transfers

### **Memory Efficiency**
- **Streaming Operations**: Memory-efficient file transfer for large files
- **Chunked Processing**: Fixed 14MB chunk size processing
- **Resource Cleanup**: Proper resource management and cleanup
- **Progress Optimization**: Pre-calculated chunk sizes for accurate progress tracking

### **Error Recovery & User Experience**
- **Graceful Error Handling**: Clear error messages with suggested solutions
- **Interactive Confirmations**: User-friendly confirmation prompts for destructive operations
- **Progress Feedback**: Real-time progress indicators with tqdm integration
- **Path Validation**: Comprehensive input validation with helpful error messages

### **SDK Usage**
```python
# Import clients
from clarifai.client.artifact import Artifact
from clarifai.client.artifact_version import ArtifactVersion

# Create artifact
artifact = Artifact().create(
    artifact_id="my-model",
    user_id="user123",
    app_id="app456",
    description="My trained model"
)

# Upload file  
version = ArtifactVersion().upload(
    file_path="./model.pt",
    artifact_id="my-model",
    user_id="user123",
    app_id="app456",
    description="Version 1.0",
    visibility="private"  # Options: private, public, org
)

# Download file
version.download(
    output_path="./downloaded_model.pt",
    artifact_id="my-model",
    version_id=version.version_id,
    user_id="user123",
    app_id="app456"
)

# List artifacts
for artifact_pb in Artifact().list(user_id="user123", app_id="app456"):
    print(f"Artifact: {artifact_pb.id}")
```

### **CLI Usage**
```bash
# List artifacts
clarifai artifact list users/user123/apps/app456
clarifai af ls users/user123/apps/app456

# Upload with description and visibility
clarifai artifact cp ./model.pt users/user123/apps/app456/artifacts/my-model \
    --description="My trained model v2.0" --visibility=org

# Download specific version
clarifai af cp users/user123/apps/app456/artifacts/my-model/versions/abc123 ./model-v2.pt

# Get artifact information
clarifai artifact get users/user123/apps/app456/artifacts/my-model

# Delete with force (no confirmation)
clarifai af delete users/user123/apps/app456/artifacts/my-model --force
```

### **Manual Testing Completed**
- ✅ All imports working correctly across different modules
- ✅ CLI commands functional with aliases (`artifact` → `af`, `list` → `ls`)
- ✅ Path parsing working for all artifact and version path formats
- ✅ Error handling graceful for edge cases and invalid inputs
- ✅ Integration with existing auth system and user context
- ✅ Progress tracking accurate for uploads and downloads
- ✅ Retry logic functioning with proper state reset

### **Automated Testing**
- ✅ 80+ test methods passing (13 artifact + 23 artifact_version + 30 CLI + 14 CLI_version tests)
- ✅ All linting checks passing with ruff
- ✅ Import tests successful across all modules
- ✅ Mock validation working for gRPC, filesystem, and network operations
- ✅ Error scenario coverage complete with comprehensive edge case testing
- ✅ Test utilities refactored to eliminate code duplication

## 🎉 **Ready for Production**

This implementation provides:

1. **✅ Complete Feature Set** - Full CRUD operations for artifacts and versions with all necessary validations
2. **✅ Production Quality** - Comprehensive testing, error handling, and network resilience
3. **✅ User Experience** - Intuitive CLI and SDK interfaces with clear feedback and error messages
4. **✅ Project Integration** - Follows all established patterns, conventions, and architectural decisions
5. **✅ Code Quality** - Refactored tests, eliminated duplication, and improved code organization
6. **✅ Documentation** - Complete with examples, usage guides, and comprehensive docstrings

**The artifact management system is ready for production deployment!** 🚀

---

## 📝 **Checklist**

- [x] SDK client implementation (Artifact + ArtifactVersion)
- [x] CLI commands with aliases (artifact/af, list/ls)
- [x] Comprehensive test suite (80+ tests across 4 test files)
- [x] Error handling and validation (UserError consistency)
- [x] Documentation and examples (comprehensive docstrings)
- [x] Code quality checks (linting, typing, ruff compliance)
- [x] Integration verification (auth system, existing patterns)
- [x] Performance optimization (streaming, chunking, progress tracking)
- [x] Network resilience features (retry logic, exponential backoff)
- [x] Test refactoring (eliminated code duplication, shared utilities)
- [x] Progress tracking improvements (accurate indexing, state reset)
- [x] Path parsing and validation (comprehensive format support)
- [x] Visibility management (private, public, org support)
- [x] Interactive CLI features (confirmations, progress bars)

### CODE REVIEW

File: clarifai/runners/artifacts/artifact_builder.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+        )
+
+
+def format_bytes(size: int) -> str:
+    """Format byte size in human readable format."""
+    for unit in ['B', 'KB', 'MB', 'GB', 'TB']:
+        if size < 1024.0:
+            return f"{size:.1f} {unit}"
+        size /= 1024.0
+    return f"{size:.1f} PB"
```
[the remark this answers is not in the document]
Comment (ydixit-clarifai) [reply clarifai/runners/artifacts/artifact_builder.py]: Done

File: tests/test_artifact.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+            assert params == expected
+
+    def test_missing_required_fields(self):
+        """Test validation with missing required fields."""
+        with patch('clarifai.client.base.BaseClient.__init__'):
+            artifact = Artifact()
+
+            # Test various missing parameter scenarios
+            with pytest.raises(UserError, match="artifact_id is required"):
+                artifact.create()
```
[the remark this answers is not in the document]
Comment (ydixit-clarifai) [reply tests/test_artifact.py]: Done

File: tests/test_artifact_version.py [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+
+        with patch('clarifai.client.base.BaseClient.__init__'):
+            version = ArtifactVersion()
+
+            # Create a mock upload config
+            upload_config = Mock()
+            upload_config.artifact_id = "test_artifact"
+
+            iterator = version._artifact_version_upload_iterator(
+                "test_file.txt", upload_config, chunk_size=4
```
[the remark this answers is not in the document]
Comment (ydixit-clarifai) [reply tests/test_artifact_version.py]: Done

File: tests/cli/test_artifact_version_cli.py [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+                    'cp',
+                    'users/test_user/apps/test_app/artifacts/test_artifact/versions/test_version',
+                    './existing_file.txt',
+                ],
+                obj=mock_obj,
+            )
+
+            assert result.exit_code != 0
+            assert "File already exists" in result.output
+            assert "File already exists" in result.output
```
[the remark this answers is not in the document]
Comment (ydixit-clarifai) [reply tests/cli/test_artifact_version_cli.py]: Updated

File: clarifai/client/artifact_version.py [hunk 5]
```diff
@@ -0,0 +1,704 @@
...
+                                version_id,
+                                user_id,
+                                app_id,
+                            ),
+                        ):
+                            if chunk_count == 0:
+                                # First response is config upload - extract version ID
+                                if hasattr(response, 'artifact_version_id'):
+                                    final_version_id = response.artifact_version_id
+                                    logger.info(f"Created artifact version: {final_version_id}")
```
[the remark this answers is not in the document]
Comment (ydixit-clarifai) [reply clarifai/client/artifact_version.py]: Updated

File: clarifai/client/artifact_version.py [hunk 6]
```diff
@@ -0,0 +1,721 @@
...
+        """
+        super().__init__(**kwargs)
+        # Ensure auth_helper is available after BaseClient init
+        if not hasattr(self, 'auth_helper'):
+            self.auth_helper = None
+        self.artifact_id = artifact_id
+        self.version_id = version_id
+        self.user_id = user_id or (
+            getattr(self.auth_helper, 'user_id', "") if self.auth_helper else ""
+        )
```
[the remark this answers is not in the document]
Comment (ydixit-clarifai) [reply clarifai/client/artifact_version.py]: Updated

File: clarifai/client/artifact_version.py [hunk 7]
```diff
@@ -0,0 +1,721 @@
...
+            except Exception as e:
+                progress_bar.close()
+                if attempt < max_retries - 1:
+                    wait_time = 2**attempt  # Exponential backoff
+                    logger.warning(
+                        f"Upload attempt {attempt + 1} failed, retrying in {wait_time}s: {e}"
+                    )
+                    import time
+
+                    time.sleep(wait_time)
```
[the remark this answers is not in the document]
Comment (ydixit-clarifai) [reply clarifai/client/artifact_version.py]: Updated

**Review by nitinbhojwani — APPROVED** [review]
LGTM
